### PR TITLE
Threat Detection new API + data masking chganges

### PR DESCRIPTION
**Pull request by yaakoviyun**

### CODE REVIEW

**Comment by markcowl** [comment]
@yaakoviyun There are merge conflicts, please update

**Comment by yaakoviyun** [comment]
Resolving conflicts.

File: src/ResourceManager/Sql/Commands.Sql/ThreatDetection/Model/BaseThreatDetectionPolicyModel.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+{
+   /// <summary>
+    /// The possible states in which an auditing policy may be in
+    /// </summary>
+    public enum ThreatDetectionStateType { Enabled, Disabled, New };
+
+    /// <summary>
+    /// The possible disable alert types
+    /// </summary> 
+    public enum DetectionType { Successful_SQLi, Attempted_SQLi, Client_GEO_Anomaly, Failed_Logins_Anomaly, Failed_Queries_Anomaly, Data_Extraction_Anomaly, Data_Alteration_Anomaly };
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/ThreatDetection/Model/BaseThreatDetectionPolicyModel.cs]: @yaakoviyun line breaks

File: src/ResourceManager/Sql/Commands.Sql/ThreatDetection/Cmdlet/SqlDatabaseThreatDetectionCmdletBase.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    /// </summary>
+    public abstract class SqlDatabaseThreatDetectionCmdletBase : AzureSqlDatabaseCmdletBase<DatabaseThreatDetectionPolicyModel, SqlThreatDetectionAdapter>
+    {
+        /// <summary>
+        /// Provides the model element that this cmdlet operates on
+        /// </summary>
+        /// <returns>A model object</returns>
+        protected override DatabaseThreatDetectionPolicyModel GetEntity()
+        {
+            return ModelAdapter.GetDatabaseThreatDetectionPolicy(ResourceGroupName, ServerName, DatabaseName, clientRequestId);
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/ThreatDetection/Cmdlet/SqlDatabaseThreatDetectionCmdletBase.cs]: line breaks

File: src/ResourceManager/Sql/Commands.Sql/ThreatDetection/Cmdlet/SetAzureSqlDatabaseThreatDetection.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        /// <returns>Returns whether the email addresses are in a correct format</returns>
+        private bool AreEmailAddressesInCorrectFormat(string emailAddresses)
+        {
+            if (string.IsNullOrEmpty(emailAddresses))
+            {
+                return true;
+            }
+
+            string[] emailAddressesArray = emailAddresses.Split(';').Where(s => !string.IsNullOrEmpty(s)).ToArray();
+            var emailRegex = new Regex(@"^(?("")("".+?(?<!\\)""@)|(([0-9a-z]((\.(?!\.))|[-!#\$%&'\*\+/=\?\^`\{\}\|~\w])*)(?<=[0-9a-z])@))(?(\[)(\[(\d{1,3}\.){3}\d{1,3}\])|(([0-9a-z][-\w]*[0-9a-z]*\.)+[a-z0-9][\-a-z0-9]{0,22}[a-z0-9]))$");
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/ThreatDetection/Cmdlet/SetAzureSqlDatabaseThreatDetection.cs]: line breaks

File: src/ResourceManager/Sql/Commands.Sql/ThreatDetection/Cmdlet/RemoveSqlDatabaseThreatDetection.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+using System.Management.Automation;
+using Microsoft.Azure.Commands.Sql.ThreatDetection.Model;
+
+namespace Microsoft.Azure.Commands.Sql.ThreatDetection.Cmdlet
+{
+    /// <summary>
+    /// Disables auditing on a specific database.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Remove, "AzureRmSqlDatabaseThreatDetectionPolicy"), OutputType(typeof(DatabaseThreatDetectionPolicyModel))]
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/ThreatDetection/Cmdlet/RemoveSqlDatabaseThreatDetection.cs]: line breaks

File: src/ResourceManager/Sql/Commands.Sql/ThreatDetection/Cmdlet/RemoveSqlDatabaseThreatDetection.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        /// Returns true if the model object that was constructed by this cmdlet should be written out
+        /// </summary>
+        /// <returns>True if the model object should be written out, False otherwise</returns>
+        protected override bool WriteResult() { return PassThru; }
+
+        /// <summary>
+        /// Updates the given model element with the cmdlet specific operation 
+        /// </summary>
+        /// <param name="model">A model object</param>
+        protected override DatabaseThreatDetectionPolicyModel ApplyUserInputToModel(DatabaseThreatDetectionPolicyModel model)
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/ThreatDetection/Cmdlet/RemoveSqlDatabaseThreatDetection.cs]: line breaks

File: src/ResourceManager/Sql/Commands.Sql/ThreatDetection/Cmdlet/GetAzureSqlDatabaseThreatDetection.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+using System.Management.Automation;
+using Microsoft.Azure.Commands.Sql.ThreatDetection.Model;
+
+namespace Microsoft.Azure.Commands.Sql.ThreatDetection.Cmdlet
+{
+    /// <summary>
+    /// Returns the auditing policy of a specific database.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Get, "AzureRmSqlDatabaseThreatDetectionPolicy"), OutputType(typeof(DatabaseThreatDetectionPolicyModel))]
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql/ThreatDetection/Cmdlet/GetAzureSqlDatabaseThreatDetection.cs]: line breaks

File: src/ResourceManager/Sql/Commands.Sql.Test/ScenarioTests/SqlTestsBase.cs [hunk 7]
```diff
@@ -157,5 +157,16 @@ protected GraphRbacManagementClient GetGraphClient()
 
             return TestBase.GetGraphServiceClient<GraphRbacManagementClient>(testFactory, tenantId);
         }
+
+        protected Management.Storage.StorageManagementClient GetStorageV2Client()
+        {
+            var client = TestBase.GetServiceClient<Management.Storage.StorageManagementClient>(new CSMTestEnvironmentFactory());
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql.Test/ScenarioTests/SqlTestsBase.cs]: line breaks

File: src/ResourceManager/Sql/Commands.Sql.Test/ScenarioTests/ThreatDetectionTests.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    public class ThreatDetectionTests : SqlTestsBase
+    {
+        protected override void SetupManagementClients()
+        {
+            var sqlCSMClient = GetSqlClient();
+            var storageClient = GetStorageClient();
+            var storageV2Client = GetStorageV2Client();
+            var resourcesClient = GetResourcesClient();
+            var authorizationClient = GetAuthorizationManagementClient();
+            helper.SetupSomeOfManagementClients(sqlCSMClient, storageClient, storageV2Client, resourcesClient, authorizationClient);
```
Comment (markcowl) [inline src/ResourceManager/Sql/Commands.Sql.Test/ScenarioTests/ThreatDetectionTests.cs]: line breaks

**Comment by markcowl** [comment]
@yaakoviyun A few minor nits, otherwise lgtm

**Pull request merged**